src/2d/shallow/bc2amr.f90

```diff
@@ -0,0 +1,280 @@
+!     can safely extrapolate in the other direction.
+!
+!     Don't overwrite ghost cells in periodic directions!!
+!     This particular routine bc2amr_noslopesets auxillary values so
+!     that no slope in topography occurs at the physical boundary.
+!
+! ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::;
+subroutine bc2amr(val,aux,nrow,ncol,meqn,naux, hx, hy, level, time,   &
+                  xlo_patch, xhi_patch, ylo_patch, yhi_patch) 
+
+    use amr_module, only: mthbc, xlower, ylower, xupper, yupper
+    use amr_module, only: xperdom,yperdom,spheredom
+
+    implicit none
+
+    ! Input/Output
+    integer, intent(in) :: nrow, ncol, meqn, naux, level
+    real(kind=8), intent(in) :: hx, hy, time
+    real(kind=8), intent(in) :: xlo_patch, xhi_patch
+    real(kind=8), intent(in) :: ylo_patch, yhi_patch
+    real(kind=8), intent(in out) :: val(meqn, nrow, ncol)
+    real(kind=8), intent(in out) :: aux(naux, nrow, ncol)
+    
+    ! Local storage
+    integer :: i, j, ibeg, jbeg, nxl, nxr, nyb, nyt
+    real(kind=8) :: hxmarg, hymarg
+
+    hxmarg = hx * .01d0
+    hymarg = hy * .01d0
+
+    ! Use periodic boundary condition specialized code only, if only one 
+    ! boundary is periodic we still proceed below
+    if (xperdom .and. (yperdom .or. spheredom)) then
+        return
+    end if
+
+    ! Each check has an initial check to ensure that the boundary is a real
+    ! boundary condition and otherwise skips the code.  Otherwise 
+    !-------------------------------------------------------
+    ! Left boundary:
+    !-------------------------------------------------------
+    if (xlo_patch < xlower-hxmarg) then
+        ! number of grid cells from this patch lying outside physical domain:
+        nxl = int((xlower + hxmarg - xlo_patch) / hx)
+
+        select case(mthbc(1))
+            case(0) ! User defined boundary condition
+                ! Replace this code with a user defined boundary condition
+                stop "A user defined boundary condition was not provided."
+            case(1) ! Zero-order extrapolation
+                do j = 1, ncol
+                    do i=1, nxl
+                        aux(:, i, j) = aux(:, nxl + 1, j)
+                        val(:, i, j) = val(:, nxl + 1, j)
+                    end do
+                end do
+
+            case(2) ! Periodic boundary condition
+                continue
+
+            case(3) ! Wall boundary conditions
+                do j = 1, ncol
+                    do i=1, nxl
+                        aux(:, i, j) = aux(:, 2 * nxl + 1 - i, j)
+                        val(:, i, j) = val(:, 2 * nxl + 1 - i, j)
+                    end do
+                end do
+                ! negate the normal velocity:
+                do j = 1, ncol
+                    do i=1, nxl
+                        val(2, i, j) = -val(2, i, j)
+                    end do
+                end do
+
+            case default
```

Comment:
We should also allow `case(4)` (sphere BCs) with a `continue` like `case(2)`, both here and for the other three boundaries.  

In principle we never get here in this case, since the module variable `spheredom` is checked earlier and should jump out of this routine, but might add clarity.

The spherical BCs haven't been tested in a while as far as I know, not sure if they still work properly...

Reply:
Added.